### PR TITLE
Implemented Replenishment SMC-ABC

### PR DESCRIPTION
**Pull request by rmjiang7**

### CODE REVIEW

File: sciope/inference/rep_smc_abc.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+        self.perturbation_kernel = perturbation_kernel
+
+        if self.use_logger:
+            self.logger = ml.SciopeLogger().get_logger()
+            self.logger.info("Sequential Monte-Carlo Approximate Bayesian Computation initialized")
+
+    def _simulate_N(self, prior, N):
+
+        params_delayed = prior.draw(N)
+        params, = dask.compute(params_delayed, num_workers = 10)
```
Comment (Wrede) [inline sciope/inference/rep_smc_abc.py]: Remove `num_workers`, let user decide the amount of workers/processes/ threads by setting up Dask them self
```suggestion
        params, = dask.compute(params_delayed)
```

File: sciope/inference/rep_smc_abc.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+            results.append(lazy_results)
+
+        computed_results, = dask.compute(results)
+
+        return np.asarray(computed_results), params
+
+    def _simulate_distance(self, param):
+        res = self.summaries_function(self.sim(param))
+        res_obs = self.summaries_function(self.data)
+        return self.distance_function.compute(res_obs, res)
```
Comment (Wrede) [inline sciope/inference/rep_smc_abc.py]: Replace by` sciope.core.get_graph_chunked()`, should support chunked data. We might need to change ` sciope.core.get_graph_chunked()` to handle a single param without a sampler function

File: sciope/inference/rep_smc_abc.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+            for _ in range(R):
+                z = self.perturbation_kernel.rvs(p)
+                num = prior.pdf(z)
+                # Metropolis Perturbation
+                if num > 0:
+                    num = num * self.perturbation_kernel.pdf(p.reshape(1,-1), z.reshape(1,-1))[0,0,]
+                    dnm = prior.pdf(p) * self.perturbation_kernel.pdf(z.reshape(1,-1), p.reshape(1,-1))[0,0]
+                    weight = num/dnm
+                    if np.random.rand() < min(1, weight):
+                        distance = self._simulate_distance(z)
```
Comment (Wrede) [inline sciope/inference/rep_smc_abc.py]: need to be relplaced if `sciope.core.get_graph_[<chunked>,<unchunked>]` are used instead

File: sciope/inference/rep_smc_abc.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+                            p = z
+                            new_distance = distance
+                            p_acc = p_acc + (1/(R * Na))
+            return p, new_distance, p_acc, n_moves
+
+        prior = self.prior_function
+        Na = round(num_samples * alpha)
+
+        # Sample N particles
+        distances, population = self._simulate_N(prior, num_samples)
```
Comment (Wrede) [inline sciope/inference/rep_smc_abc.py]: `sciope.core`

File: sciope/inference/rep_smc_abc.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+    * summaries_function    	(summary statistics calculation function)
+    * distance_function         (function calculating deviation between simulated statistics and observed statistics)
+    * summaries_divisor         (numpy array of maxima - used for normalizing summary statistic values)
+    * use_logger    			(whether logging is enabled or disabled)
+
+    Methods:
+    * infer 					(perform parameter inference)
+    """
+
+    def __init__(self, data, sim, prior_function, perturbation_kernel,
```
Comment (Wrede) [inline sciope/inference/rep_smc_abc.py]: lack numpy style docstring

File: sciope/inference/smc_abc.py [hunk 6]
```diff
@@ -92,7 +92,7 @@ class SMCABC(InferenceBase):
     """
 
     def __init__(self, data, sim, prior_function, perturbation_kernel,
```
Comment (Wrede) [inline sciope/inference/smc_abc.py]: lack numpy style docstring, or does it inherits from `InferenceBase`?

File: sciope/utilities/epsilonselectors/epsilon_selector.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
+from abc import ABCMeta, abstractmethod
+
+class EpsilonSelector(object):
```
Comment (Wrede) [inline sciope/utilities/epsilonselectors/epsilon_selector.py:7]: lack class doc/description

File: sciope/utilities/epsilonselectors/absolute_epsilon_selector.py [hunk 8]
```diff
@@ -0,0 +1,15 @@
+from sciope.utilities.epsilonselectors.epsilon_selector import *
+
+class AbsoluteEpsilonSelector(EpsilonSelector):
```
Comment (Wrede) [inline sciope/utilities/epsilonselectors/absolute_epsilon_selector.py:7]: lack class doc/description

File: sciope/utilities/epsilonselectors/relative_epsilon_selector.py [hunk 9]
```diff
@@ -0,0 +1,31 @@
+from sciope.utilities.epsilonselectors.epsilon_selector import *
+import numpy as np
+
+class RelativeEpsilonSelector(EpsilonSelector):
```
Comment (Wrede) [inline sciope/utilities/epsilonselectors/relative_epsilon_selector.py:4]: lack class doc/description

**Review by Wrede — COMMENTED** [review]
In general we should try to make use of sciope.core to make it consistent and fault tolerant. Will SMC-ABC be able to handle chunked computations on worker nodes?

**Review by prasi372 — APPROVED** [review]

**Pull request merged**